invokeai/app/invocations/prompt_tools.py

```diff
@@ -0,0 +1,286 @@
+class PromptJoinInvocation(BaseInvocation):
+    """Joins prompt left to prompt right"""
+
+    prompt_left: str = InputField(default="", description="Prompt Left", ui_component=UIComponent.Textarea)
+    prompt_right: str = InputField(default="", description="Prompt Right", ui_component=UIComponent.Textarea)
+
+    def invoke(self, context: InvocationContext) -> StringOutput:
+        return StringOutput(value=((self.prompt_left or "") + (self.prompt_right or "")))
+
+
+@invocation("prompt_join_three", title="Prompt Join Three", tags=["prompt", "join"], category="prompt", version="1.0.0")
+class PromptJoinThreeInvocation(BaseInvocation):
+    """Joins prompt left to prompt middle to prompt right"""
+
+    prompt_left: str = InputField(default="", description="Prompt Left", ui_component=UIComponent.Textarea)
+    prompt_middle: str = InputField(default="", description="Prompt Middle", ui_component=UIComponent.Textarea)
+    prompt_right: str = InputField(default="", description="Prompt Right", ui_component=UIComponent.Textarea)
+
+    def invoke(self, context: InvocationContext) -> StringOutput:
+        return StringOutput(value=((self.prompt_left or "") + (self.prompt_middle or "") + (self.prompt_right or "")))
+
+
+@invocation(
+    "prompt_replace", title="Prompt Replace", tags=["prompt", "replace", "regex"], category="prompt", version="1.0.0"
+)
+class PromptReplaceInvocation(BaseInvocation):
+    """Replaces the search string with the replace string in the prompt"""
+
+    prompt: str = InputField(default="", description="Prompt to work on", ui_component=UIComponent.Textarea)
+    search_string: str = InputField(default="", description="String to search for", ui_component=UIComponent.Textarea)
+    replace_string: str = InputField(
+        default="", description="String to replace the search", ui_component=UIComponent.Textarea
+    )
+    use_regex: bool = InputField(
+        default=False, description="Use search string as a regex expression (non regex is case insensitive)"
+    )
+
+    def invoke(self, context: InvocationContext) -> StringOutput:
+        pattern = self.search_string or ""
+        new_prompt = self.prompt or ""
+        if len(pattern) > 0:
+            if not self.use_regex:
+                # None regex so make case insensitve
+                pattern = "(?i)" + re.escape(pattern)
+            new_prompt = re.sub(pattern, (self.replace_string or ""), new_prompt)
+        return StringOutput(value=new_prompt)
+
+
+class PTFields(BaseModel):
+    """Prompt Tools Fields for an image generated in InvokeAI."""
+
+    positive_prompt: str
+    positive_style_prompt: str
+    negative_prompt: str
+    negative_style_prompt: str
+    seed: int
+    width: int
+    height: int
+    steps: int
+    cfg_scale: float
+
+
+@invocation_output("pt_fields_collect_output")
+class PTFieldsCollectOutput(BaseInvocationOutput):
+    """PTFieldsCollect Output"""
+
+    pt_fields: str = OutputField(description="PTFields in Json Format")
```

Comment:
Can you please explain the use case for the `PTFields` model? It seems to be used as intermediate storage of a select number of fields. Why not use a primitive for each? I'm wary of using a new data structure instead of composing existing ones

Reply:
See Previous comment as I think I answer this there.